### PR TITLE
Implement Face Detection

### LINKED ISSUE #27: [ros2-devel] Face Detection
The ROS2 ADA feeding system needs face detection. This issue encompasses the following:

1. Read and understand the [ROS1 face detection code](https://github.com/personalrobotics/face_detection/tree/master) (in C++) we used.
2. Familiarize yourself with the desired interface for ROS2 face detection, which consists of a ROS service to toggle face detection on and off, and while it is on it publishes results to a ROS topic. The best way to familarize yourself with this is to look at the [face detection dummy node](https://github.com/personalrobotics/feeding_web_interface/blob/main/feeding_web_app_ros2_test/feeding_web_app_ros2_test/FaceDetection.py).
3. Investigate ways to do face detection in Python. We not only need to detect the stomion (mouth center) in the 2D image, but we also have to translate it into a 3D point relative to the camera frame.
4. Implement face detection in Python.

### PR DESCRIPTION
**Pull request by taylorkf**
# Description

This PR adds a face detection node, addressing https://github.com/personalrobotics/ada_feeding/issues/27. It also addresses some issues with segment_from_point (python 3 typing errors and a function call that was possibly blocking realsense nodes).

# Testing procedure

On the realsense camera (using t0b1/nano), I tested that the node starts and stops with the appropriate service call, and publishes mouth locations only when a face is detected.

**Initialization**:
1. SSH into nano@nano
2. On nano, start vision nodes:
    1. `ros2config`
    2. `ros2 launch realsense2_camera rs_launch.py rgb_camera.profile:='640,480,30' depth_module.profile:='640,480,30' align_depth.enable:='true' initial_reset:='true'`
3. On t0b1, launch the face_detection node:
    1. `ros2 launch ada_feeding_perception ada_feeding_perception_launch.xml`
4. Listen to the /face_detection topic:
    1. `ros2 topic echo /face_detection`

**Testing**:
1. For each of the following commands, run it and verify that it executes as expected. 
    1. Turn on face detection 
        - [x] `ros2 service call /toggle_face_detection std_srvs/srv/SetBool "{data: True}"`
        - [x] The face detection node should receive this message
   2. Turn off face detection
        - [x]  `ros2 service call /toggle_face_detection std_srvs/srv/SetBool "{data: False}"`
        - [x] The face detection node should receive this message
2. Position your face in front of the camera and move to several different positions. Make sure that the face_detection topic returns the correct values.
     

# Before opening a pull request
- [x] Format your code using [black formatter](https://black.readthedocs.io/en/stable/) `python3 -m black .`
- [x] Run your code through [pylint](https://pylint.readthedocs.io/en/latest/) and address all warnings/errors. The only warnings that are acceptable to not address is TODOs that should be addressed in a future PR. From the top-level `ada_feeding` directory, run: `pylint --recursive=y --rcfile=.pylintrc .`.

# Before Merging
- [ ] `Squash & Merge`

### CODE REVIEW

File: ada_feeding_perception/package.xml
Comment (amalnanavati) [inline ada_feeding_perception/package.xml:14]: Fascinating that this had to be a `test_depend` and nothing else. 

No need to change this, I just find it interesting

File: ada_feeding_perception/ada_feeding_perception/face_detection.py
Comment (amalnanavati) [inline ada_feeding_perception/ada_feeding_perception/face_detection.py:285]: Is there a website you can link to that says what each landmark corresponds to? That would be helpful for future folks to understand this code.

File: ada_feeding_perception/ada_feeding_perception/face_detection.py
Comment (amalnanavati) [inline ada_feeding_perception/ada_feeding_perception/face_detection.py:274]: I'm really confused by this loop over `landmarks`. You passed into `landmark_detector.fit` a list of faces. So presumably landmarks is a 2D list where the first dimension corresponds to faces, and the second dimension corresponds to the landmarks within that face, right?

Then, why are you looping over landmarks, and then looping over faces? Isn't that essentially looping twice over faces? You also don't refer to the variable `landmark`. Further, it seems like the main functional part of this loop is annotating `cv_image`. But (1) you are re-generating that variable for every landmark; and (b) you are re-creating a variable you already created above.

I'm suggest the following:
1. At the top of the function, create two CV images, one to annotate and one grayscale to pass into the detectors.
2. Move the `for face in faces` logic after faces have been detected. It doesn't seem like it actually has anything to do with the landmarks.

File: ada_feeding_perception/ada_feeding_perception/face_detection.py
Comment (amalnanavati) [inline ada_feeding_perception/ada_feeding_perception/face_detection.py:286]: You refer to `landmark` here, but this is outside the loop `for landmark in landmarks` which is confusing and seems rife for a bug.

File: ada_feeding_perception/ada_feeding_perception/face_detection.py
Comment (amalnanavati) [inline ada_feeding_perception/ada_feeding_perception/face_detection.py:273]: Can you add a comment here explaining the structure of `landmarks` (e.g., is it a 2D list, a 3D list, etc. and what does each dimension represent?)

File: ada_feeding_perception/ada_feeding_perception/face_detection.py
Comment (amalnanavati) [inline ada_feeding_perception/ada_feeding_perception/face_detection.py:291]: `annotated_img` actually seems like a misnamed variable since you have been making the distinction between `images` and `msgs`.

Why not remove the line `annotated_img = annotated_msg` and then change like 297 (annotated_img = msg) to `annotated_msg = msg` and then publish `annotated_msg`? That seems like a clearer variable name, and then prevents creating two variables (annotated_msg and annotated_img) that contain the same data.

File: ada_feeding_perception/ada_feeding_perception/face_detection.py
Comment (amalnanavati) [inline ada_feeding_perception/ada_feeding_perception/face_detection.py:28]: Technically, for best coding practices you should have a class docstring as well. Can you add one below the class definition? Feel free to have it be mostly the same as your module docstring at the top of the file.

File: ada_feeding_perception/ada_feeding_perception/face_detection.py
Comment (amalnanavati) [inline ada_feeding_perception/ada_feeding_perception/face_detection.py:103]: I don't think this is necessary (neither is the one below). Because they are class attributes (e.g., `self.`) the linter won't give an unused variable warning here.

File: ada_feeding_perception/ada_feeding_perception/face_detection.py
Comment (amalnanavati) [inline ada_feeding_perception/ada_feeding_perception/face_detection.py:187]: Please add type annotations here. I think it should be:

`def toggle_face_detection_callback(self, request: ToggleFaceDetection.Request, response: ToggleFaceDetection.Response) -> ToggleFaceDetection.Response:`

File: ada_feeding_perception/ada_feeding_perception/face_detection.py
Comment (amalnanavati) [inline ada_feeding_perception/ada_feeding_perception/face_detection.py:208]: Type annotations

File: ada_feeding_perception/ada_feeding_perception/face_detection.py
Comment (amalnanavati) [inline ada_feeding_perception/ada_feeding_perception/face_detection.py:241]: Type annotations

File: ada_feeding_perception/package.xml
Comment (amalnanavati) [inline ada_feeding_perception/package.xml:1]: Could you add the multiple maintainer's emails to this file as well, as you did in `setup.py`?

File: ada_feeding_perception/launch/ada_feeding_perception_launch.xml
Comment (amalnanavati) [inline ada_feeding_perception/launch/ada_feeding_perception_launch.xml:2]: I don't like the idea of pushing a change where we are removing food detection. I think it was fine for testing face detection, but we should identify the issue before merging the changes in, so that face detection and food detection can both run at the same time.

File: ada_feeding_perception/ada_feeding_perception/face_detection.py
Comment (amalnanavati) [inline ada_feeding_perception/ada_feeding_perception/face_detection.py:233]: The depth image coordinates are in `mm`. However, the norm in ROS is for `PoseStamped` points to be in `m`. Can you convert them before putting them into the message?

**Review by amalnanavati — CHANGES_REQUESTED** [review]
See in-line comments. There is one major bug re. converting 2D pixels to 2D points, some code that seems unnecessary / doesn't make sense regarding iterating over landmarks, and most of the rest of the in-line comments are small.

Additionally, I have four broader points of discussion:
1. It seems like your code finds the stomion of the mouth in the RGB image and then gets its depth in the depth image. But if the mouth is open, wouldn't the depth of the stomion be really far into the mouth/throat? I think we should use the (u,v) position of the stomion, but for the depth we should average (or min or max) all the detected landmarks along the lips.
2. After detecting a mouth center, it looks like you get the depth of the mouth on the *next* depth image (as opposed to on the *closest* one). In practice, at what rate do depth images come in relative to the RGB images; is it actually the case that the next one will be the closest one? Because if we are using face detection during visual servoing (i.e., while the robot arm is moving) then even a 0.25 sec delay between those images could result in the (u,v) coordinates that were the mouth center now being on the nose, or to the side of the face, resulting in wrong depth measurements.
    1. To address this, we might need to store depth images e.g., from the last second, and find the closest depth image to the RGB one, not just the next depth image.
3. What if multiple faces are detected? (e.g., Tyler's caregiver is standing behind Tyler, or a fascinated friend is staring at the robot). Firstly, what would the detector do in that case? Secondly, I think what it *should do* is have an "expected face location" (e.g., near wheelchair headrest) in a fixed world frame that is passed in as a parameter, and it selects the face that is closest to that point.
4. I'm concerned that our testing procedure is not rigorous enough, given that it overlooked the bug re. the conversion from pixels to 3D coordinates. Can you add an additional test, where you run the face detector, move your face around, and ensure that the detected point actually aligns with your real mouth's distance from the camera? Try your head in several different poses, and the robot arm in several different poses.

**Comment by taylorkf** [comment]
@amalnanavati Thanks for the bug catches! My lack of experience working with cv is showing 😆 Actually writing down a quick algorithm this time and will get this coded up on Monday when I'm back in town.

File: ada_feeding_perception/ada_feeding_perception/face_detection.py
Comment (amalnanavati) [inline ada_feeding_perception/ada_feeding_perception/face_detection.py:95]:
```suggestion
            "toggle_face_detection",
```
[ROS naming conventions](http://wiki.ros.org/ROS/Patterns/Conventions) have topics and services be in snake_case. (I know this was a carryover from the dummy node -- that was my mistake)

File: ada_feeding_perception/ada_feeding_perception/face_detection.py
Comment (amalnanavati) [inline ada_feeding_perception/ada_feeding_perception/face_detection.py]: nit: Largest detected _mouth_ or _face_?

File: ada_feeding_perception/ada_feeding_perception/face_detection.py
Comment (amalnanavati) [inline ada_feeding_perception/ada_feeding_perception/face_detection.py:41]: In line with our conversation with Ethan, can you add a TODO here that eventually this node should return all detected faces (perhaps with an ID if this node is tracking faces) and then let the client decide which to use? (As opposed to what currently happens where this node decides which of multiple faces to return, based on the size in pixel space.)

File: ada_feeding_perception/ada_feeding_perception/face_detection.py
Comment (amalnanavati) [inline ada_feeding_perception/ada_feeding_perception/face_detection.py]: It looks like both `face_model_name` and `landmark_model_name` are only used in the `__init__` function and therefore don't need to be class attributes. If that is truly the case, can you just make them local variables instead?

File: ada_feeding_perception/ada_feeding_perception/face_detection.py
Comment (amalnanavati) [inline ada_feeding_perception/ada_feeding_perception/face_detection.py:98]: Nit: Can you initialize these with the service that changes these values (i.e., on line 117) to help with readability?

File: ada_feeding_perception/ada_feeding_perception/face_detection.py
Comment (amalnanavati) [inline ada_feeding_perception/ada_feeding_perception/face_detection.py]: Same here -- if you could define the depth buffer along with the depth subscriber (before, but still in adjacent lines of code) that would help me intuit what its purpose is. Currently, I had to Ctrl+F and look further in the code to understand what this variable's purpose is.

File: ada_feeding_perception/ada_feeding_perception/face_detection.py
Comment (amalnanavati) [inline ada_feeding_perception/ada_feeding_perception/face_detection.py]: Which img is this? The latest img received or something else? The color image or depth image? At a minimum a comment would help here, but also moving this line to be near the subscriber it is used for will help with readability.

File: ada_feeding_perception/ada_feeding_perception/face_detection.py
Comment (amalnanavati) [inline ada_feeding_perception/ada_feeding_perception/face_detection.py]: Can you specify _which_ values these are? It looks like it is the intrinsics camera matrix (the [K matrix in CameraInfo](http://docs.ros.org/en/melodic/api/sensor_msgs/html/msg/CameraInfo.html))

File: ada_feeding_perception/ada_feeding_perception/face_detection.py
Comment (amalnanavati) [inline ada_feeding_perception/ada_feeding_perception/face_detection.py]: The way I [use this in `MoveToMouth`](https://github.com/personalrobotics/ada_feeding/blob/a6a6655cfb5b6a05a91a120b906175be2cb9fcb7/ada_feeding/ada_feeding/trees/move_to_mouth_tree.py#L149), I expect the result to indicate whether toggling succeeded or not. In other words, if the request was `False` and you successfully turned it off, the response should be `{'success': True}`. Also, can you also populate [the `message` field of SetBool](http://docs.ros.org/en/noetic/api/std_srvs/html/srv/SetBool.html)?

I think the best way to do this is:
```
response.success = False
response.message = f"Failed to set is_on to {request.data}"
with self.is_on_lock:
    self.is_on = request.data
    response.success = True
    response.message = f"Successfully set is_on to {request.data}"
return response
```

File: ada_feeding_perception/ada_feeding_perception/face_detection.py
Comment (amalnanavati) [inline ada_feeding_perception/ada_feeding_perception/face_detection.py:276]: Since `self.camera_matrix` can be accessed across multiple threads, it should have a lock. It looks like lines 294-298 in `depth_callback` below use it. I'd suggest creating a local copy of the variable at the top of `depth_callback`, like you do with `self.is_on`/`self.is_on_lock`, and then using that local copy (not the instance attribute attribute) throughout the function

File: ada_feeding_perception/ada_feeding_perception/face_detection.py
Comment (amalnanavati) [inline ada_feeding_perception/ada_feeding_perception/face_detection.py:275]: Can you add the following in this docstring?

```
TODO: We technically only need to read one message from CameraInfo, since the intrinsics don't change. If `rclpy` adds a `wait_for_message` function, this subscription/callback should be replaced with that.
```

File: ada_feeding_perception/ada_feeding_perception/face_detection.py
Comment (amalnanavati) [inline ada_feeding_perception/ada_feeding_perception/face_detection.py]: Also, can you make the length of `self.depth_buffer` a parameter with default value 10? Because if in the future we increase the FPS of the depth camera, the number of images stored in the buffer will also need to be increased proportionally.

File: ada_feeding_perception/ada_feeding_perception/face_detection.py
Comment (amalnanavati) [inline ada_feeding_perception/ada_feeding_perception/face_detection.py]: I think `self.is_face_detected` should also be protected by a lock, since it can be accessed across multiple threads. Same with the other attributes that are set in `color_callback` and used here: `self.img_mouth_center`, `self.img_timestamp`, and `self.img_mouth_points`. The main reason for this is to ensure that all these attributes are read from a message of the same timestamp. As it stands, if a call to `color_callback` is being run at the same time as this `depth_callback`, you might get the `img_mouth_center` from the previous call but the `img_mouth_points` from the new call, which is undesirable.

Comment (amalnanavati) [reply ada_feeding_perception/ada_feeding_perception/face_detection.py]: To address this, I'd suggest having one lock `self.mouth_detection_lock` to handle all instance attributes that are set in `color_callback`. In `color_callback`, do all your computations with local variables, and only at the end of the function you should acquire the lock and set all the instance attributes. Similarly, at the top of `depth_callback` you should acquire the lock, store all instance attributes as local variables, and then only use the local variables for the rest of the function.

File: ada_feeding_perception/ada_feeding_perception/face_detection.py
Comment (amalnanavati) [inline ada_feeding_perception/ada_feeding_perception/face_detection.py]: Nit: since you know the length of `difference_array` before-hand (the same length as `depth_img`) you should instead create an array of that size (e.g., `difference_array = np.zeros((len(self.depth_buffer),), stype=np.float32)`) and then iterate over the indices (`for i in range(len(self.depth_buffer)`) and set that index of the difference array instead of appending (`difference_array[i] = abs(depth_time - img_time)`). This prevents multiple changes of the size of the underlying array.

File: ada_feeding_perception/ada_feeding_perception/face_detection.py
Comment (amalnanavati) [inline ada_feeding_perception/ada_feeding_perception/face_detection.py:336]: Can you add a link to the YouTube video you got these formulae from, so a reader of the code can refer back to that if need be?

File: ada_feeding_perception/ada_feeding_perception/face_detection.py
Comment (amalnanavati) [inline ada_feeding_perception/ada_feeding_perception/face_detection.py]: Looks like this line isn't used unless `is_on` is True -- therefore, can you move it to inside that conditional?

File: ada_feeding_perception/ada_feeding_perception/face_detection.py
Comment (amalnanavati) [inline ada_feeding_perception/ada_feeding_perception/face_detection.py]: JW, did you do an `imshow` after loading `image_rgb` to check whether it is actually RGB or BGR? OpenCV's default way to store images is BGR whereas ROS's is RGB. You should do:

```
cv2.imshow("image_rgb", image_rgb)
cv2.waitKey(0)
```

If that image's colors appear correctly, that means it is BGR, which means the above line should be `cv2.COLOR_BGR2GRAY`. If that image's colors appear incorrectly, that means it is RGB, which means the above line is correct.

File: ada_feeding_perception/ada_feeding_perception/face_detection.py
Comment (amalnanavati) [inline ada_feeding_perception/ada_feeding_perception/face_detection.py]: I'd actually prefer if the header is the image message header as opposed to the depth message header, so that there is a one-to-one correspondence between the headers in the annotated image message and the FaceDetection message.  Since you already have the img_timestamp, it is just a matter of `mouth_location.header = Header(stamp=img_timestamp)`

File: ada_feeding_perception/ada_feeding_perception/face_detection.py
Comment (amalnanavati) [inline ada_feeding_perception/ada_feeding_perception/face_detection.py]: Just wondering, does this need to be re-created for every image message, or can you create it once in `__init__` and reuse it for every image message? (The latter is more efficient if it works.)

File: ada_feeding_perception/ada_feeding_perception/face_detection.py
Comment (amalnanavati) [inline ada_feeding_perception/ada_feeding_perception/face_detection.py]: Also wondering if this can be initialized/loaded just once in `__init__` instead of every time in the callback. Loading the models probably slows the callback down quite a bit.

File: ada_feeding_perception/ada_feeding_perception/face_detection.py
Comment (amalnanavati) [inline ada_feeding_perception/ada_feeding_perception/face_detection.py]:
```suggestion
                largest_face = [0, 0] # [area (px^2), index]
```

File: ada_feeding_perception/ada_feeding_perception/face_detection.py
Comment (amalnanavati) [inline ada_feeding_perception/ada_feeding_perception/face_detection.py]:
```suggestion
                    for j in range(48, 68):
```

Python is exclusive of the upper bound.

File: ada_feeding_perception/ada_feeding_perception/face_detection.py
Comment (amalnanavati) [inline ada_feeding_perception/ada_feeding_perception/face_detection.py]:
```suggestion
                self.img_mouth_points = landmarks[largest_face[1]][0][48:68]
```

File: ada_feeding_perception/ada_feeding_perception/helpers.py
Comment (amalnanavati) [inline ada_feeding_perception/ada_feeding_perception/helpers.py]: I'm surprised that this `np.ndarray` didn't cause an issue. Anyway, no need to change it as it is getting fixed in #47

File: ada_feeding_perception/config/face_detection.yaml
Comment (amalnanavati) [inline ada_feeding_perception/config/face_detection.yaml]: Nit: I think convention is to have just a single newline at the bottom of every file (not multiple).

**Review by amalnanavati — CHANGES_REQUESTED** [review]
Overall it looks good! Most of the below comments are nits to help with readability. The only functional changes are:
1. Changing how `toggle_face_detection` response is handled.
2. Adding locks to make this node thread-safe.
3. (If possible) move model loading to `__init__`.
4. Fixing indexing for mouth landmarks.

File: ada_feeding_perception/ada_feeding_perception/face_detection.py
Comment (amalnanavati) [reply ada_feeding_perception/ada_feeding_perception/face_detection.py:98]: In general, variables that are used in similar ways should be initialized on nearby lines of code.

File: ada_feeding_perception/ada_feeding_perception/face_detection.py
Comment (taylorkf) [reply ada_feeding_perception/ada_feeding_perception/face_detection.py]: Since I'm not in to check this today, I'm changing this to BGR2GRAY. Can you check this? Or I can check it once I'm not sick.

**Comment by amalnanavati** [comment]
Tested it, made some slight changes but other than that it looks good.

I'm merging it in, but as a side-note I think the fact that camera subscriptions is unreliable will be a big problem for us. For example, if while running `ada_feeding_perception_launch.xml`, I also use the commond-line tool to check the hz of the aligned depth topic, that is one subscription too many and the face detection node stops receiving color images. We do need an in-depth investigation into when the problem arises, and whether a republisher (started in #49 ) can address it, and if so what the minimal set of topics to republish is.

**Review by amalnanavati — APPROVED** [review]

File: ada_feeding_perception/ada_feeding_perception/face_detection.py
Comment (amalnanavati) [reply ada_feeding_perception/ada_feeding_perception/face_detection.py]: Verified that BGR2GRAT is correct

**Pull request merged**
